New version of asyn fix

MarkRivers opened the pull request:
This version is based on the new version of devAsynInt32.c from asyn.  It has many deletions from that file, and replaces the bo record with the busy record.

MarkRivers committed:
New version that is based on the devAsynInt32.c from asyn.
It uses a ring buffer and the callbackRequest() call rather than
scanOnce to ensure that the driver is not called when the record
processes due to a driver callback.

MarkRivers committed:
New version of asyn

MarkRivers committed:
Fix comments

MarkRivers commented:
This version also adds addtional features that were added to devAsynInt32.c since the previous version of devBusyAsyn.c was created.
- There is a ring buffer for callback values, and the size of this can  be controlled with the info tag asyn:FIFO.  The default size is 10.
- The driver can use the pasynUser fields auxStatus, timeStamp, alarmStatus and alarmSeverity to set the corresponding values in the record.  
- The driver can change the enumStrings from those those in the database.
- If the driver is disconnected or returns an error there will only be a single error message on the disconnect or error, and a single message when it reconnects or changes error status.  This prevents large numbers of error messages,

anjohnson commented on busyApp/src/devBusyAsyn.c:
```diff
@@ -35,38 +37,79 @@
 #include <recSup.h>
 #include <devSup.h>
 
+#include <epicsExport.h>
```
This doesn't look like the right place for the `#include <epicsExport.h>` to me. This line needs to go below the imports of code from other libraries (in this case Asyn is an external library) and above includes of other headers that declare code in the same library. The busy record itself is presumably in the same library as this device support, although since device support doesn't usually call routines that get declared in the record.h file it doesn't really matter where that header gets included. The previous location (below all the other includes) is thus OK.

MarkRivers replied:
Thanks for catching that.  This was the result of a cut and paste error, because in devAsynInt32.c epicsExport.h needs to be before the asyn include files, since they are in the same DLL.  But in busy epicsExport.h needs to be after the asyn include files.  I also moved busyRecord.h after epicsExport.h for consistency.

anjohnson commented on busyApp/src/devBusyAsyn.c:
```diff
@@ -35,38 +37,79 @@
+#define INIT_ERROR -1
+
+#define DEFAULT_RING_BUFFER_SIZE 10
+/* We should be getting these from db_access.h, but get errors including that file? */
```
You can never include db_access.h and dbAccess.h in the same source file because they reuse names such as the DBF_* types; the former is for CA clients, the latter for IOC code. The two `#define`s on the line below this comment are giving CA's limits for enums, which are not quite the same as the IOC's limits (the latter may be more relevant for pvAccess users in the future). The dbAccess code allows up to `DB_MAX_CHOICES` (30, defined in dbDefs.h) states with names of up to `MAX_STRING_SIZE` (40, defined in epicsTypes.h) characters. For compatibility the mbbi record will probably always implement the CA limits (there was a time when it only allowed even shorter strings), but there might be record types in the future that allow for more states and longer strings.

MarkRivers replied:
Thanks for pointing that out.  I have changed to use DB_MAX_CHOICES and MAX_STRING_SIZE to allow for future record types.  I will do the same thing in asyn device support.

anjohnson reviewed (approved):
I haven't been through all the code changes in detail, there are just a couple of comments below, but in general the idea seems good to me.

- Andrew

MarkRivers committed:
Fix location of epicsExport.h.
Change MAX_ENUM_STATES to DB_MAX_CHOICES and MAX_ENUM_STRING_SIZE to MAX_STRING_SIZE.
The previous macros are for CA clients and defined in db_access.h,
the new ones are for IOC code and are in dbAccess.h.

MarkRivers committed:
Fixes to compile on Windows; changed devInt32Pvt to devPvt for consistency with standard asyn device support

MarkRivers commented:
I made some minor changes to compile on Windows because there were declarations after the first executable statement.  I also changed devInt32Pvt to devPvt for consistency with the files in standard asyn device support.

MarkRivers committed:
Revert changing MAX_ENUM_STATES to DB_MAX_CHOICES and MAX_ENUM_STRING_SIZE to MAX_STRING_SIZE, former was correct.
Fix callbacks for output records; call dbProcess, not prset->process; pop element from ring buffer if dbProcess did not process record.

MarkRivers committed:
Update for R1-7
